db/log_reader.cc

```diff
@@ -68,7 +69,8 @@ Reader::~Reader() {
 // restrict the inconsistency to only the last log
 bool Reader::ReadRecord(Slice* record, std::string* scratch,
                         WALRecoveryMode wal_recovery_mode,
-                        uint64_t* record_checksum) {
+                        uint64_t* record_checksum,
+                        std::map<uint32_t, size_t>* cf_to_ts_sz) {
```

Comment:
Thanks for the suggestion to use `unordered_map`, we indeed don't need to sorting overhead. 

It's a good idea to add a separate API for timestamp size related read/writing, especially because the existing APIs are used by both WAL logs and manifest files. I started with making it an additional argument in the existing API thinking that the log writter can internally enforce the timestamp size record be added prior to the WriteBatch record. But since there is only one place in db_impl_write where WAL log is written, I can just make sure to log the timestamp size before log the WriteBatch.

The log reader is also updated to have a similar change to make it a separate API. But instead of making the API return timestamp size for a given column family, it just returns the accumulated timestamp size records it has read so far. I'm thinking that later, when replay a WriteBatch in the log, I can do something like collect column families involved in a WriteBatch, check if these column families potentially have a discrepancy before doing a best effort reconciliation. In that usage, having the whole map would make the implementation more straightforward. WDYT?

Reply:
SGTM!